plugin-server/src/cdp/services/job-queue/job-queue-delay.test.ts

```diff
@@ -0,0 +1,110 @@
+    })
+
+    it('routes immediately to returnTopic when scheduled in the past', async () => {
+        const headers = createHeaders({
+            returnTopic: 'cdp_cyclotron_hog',
+            queueScheduledAt: new Date(Date.now() - 1000).toISOString(),
+        })
+        const msg = createKafkaMessage({ headers })
+
+        const p = (queue as any)['consumeKafkaBatch']([msg])
+        await Promise.resolve()
+        jest.advanceTimersByTime(0)
+        await p
+
+        expect(mockProducer.produce).toHaveBeenCalledWith(
+            expect.objectContaining({
+                value: msg.value,
+                key: msg.key as any,
+                topic: 'cdp_cyclotron_hog',
+                headers: headers as any,
+            })
+        )
+        expect(mockConsumer.offsetsStore).toHaveBeenCalledWith([{ ...msg, offset: msg.offset + 1 }])
+        expect(consumeBatch).toHaveBeenCalledWith([])
+    })
+
+    it('waits up to scheduled time (short wait) then routes to returnTopic', async () => {
+        const waitMs = 5000
+        const headers = createHeaders({
+            returnTopic: 'cdp_cyclotron_hog',
+            queueScheduledAt: new Date(Date.now() + waitMs).toISOString(),
+        })
+        const msg = createKafkaMessage({ headers })
+
+        const promise = (queue as any)['consumeKafkaBatch']([msg])
+        jest.advanceTimersByTime(waitMs)
+        await Promise.resolve()
+        await promise
+
+        expect(mockProducer.produce).toHaveBeenCalledWith(expect.objectContaining({ topic: 'cdp_cyclotron_hog' }))
+        expect(mockConsumer.offsetsStore).toHaveBeenCalledWith([{ ...msg, offset: msg.offset + 1 }])
+    })
+
+    it('caps wait at 10m and re-queues to delay topic when still in the future', async () => {
+        const longMs = getDelayByQueue('delay_10m') * 3
+        const tenMinutes = getDelayByQueue('delay_10m')
+        const headers = createHeaders({
+            returnTopic: 'cdp_cyclotron_hog',
+            queueScheduledAt: new Date(Date.now() + longMs).toISOString(),
+        })
+        const msg = createKafkaMessage({ headers })
+
+        const promise = (queue as any)['consumeKafkaBatch']([msg])
+        jest.advanceTimersByTime(tenMinutes)
+        await Promise.resolve()
+        await promise
+
+        expect(mockProducer.produce).toHaveBeenCalledWith(
+            expect.objectContaining({
+                topic: 'cdp_cyclotron_delay_10m',
+            })
+        )
+        expect(mockConsumer.offsetsStore).toHaveBeenCalledWith([{ ...msg, offset: msg.offset + 1 }])
+    })
+
+    it('skips message and commits offset when required headers are missing', async () => {
+        const msgNoHeaders = createKafkaMessage({ headers: [] })
+
+        await (queue as any)['consumeKafkaBatch']([msgNoHeaders])
+
+        expect(mockProducer.produce).not.toHaveBeenCalled()
+        expect(mockConsumer.offsetsStore).toHaveBeenCalledWith([{ ...msgNoHeaders, offset: msgNoHeaders.offset + 1 }])
+    })
+})
```

Comment:
I think it could be good to have a test with a long timeout that ensures the shutdown happens cleanly and isnt blocked by it